Mapping CSV Tool and EOLS General Transformations Code

quazi-h opened the pull request:
## Why

[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1808)

## This PR
Added a mapping transformation script to process schema csvs to json schemas and fragments
Added EOLS Fragments and general conditions to the dog aging schema.
Added EOLS pipeline, pipeline builder, transformation code, and unit tests for general conditions.

aherbst-broad committed:
Add mapping transformation script

aherbst-broad committed:
Fix

aherbst-broad committed:
Fix primary key

quazi-h committed:
Merge branch 'master' into qh-arh-dspdc-1808-convert-tool

quazi-h committed:
Updates to the schema exporter tool to add a primary key in the correct position.
EOLS Fragments and general conditions.
EOLS pipeline builders and transformation code for general conditions.

quazi-h committed:
Changing the output of the mapEols functiont to be an Eols object instead of an option.
Created a unit test for the general Eols conditions.
Realized that the RedCap data dict has some gating logic for some of those fields, added notes on what needs to be refactored

quazi-broad commented:
Realized that the RedCap data dict has some gating logic for some of those fields, added notes on what needs to be refactored.

aherbst-broad commented on scripts/convert_mapping_to_schema.py:
```diff
@@ -0,0 +1,117 @@
+
+
+def render_schemas(fragments: dict[str, list[JadeColumn]], table_name: str):
+    general_fragment = fragments.pop(TDR_FRAGMENT_GENERAL)
```
🔧 I think this should handle the edge case where there is no general fragment ( i.e., everything is bucketed) with a default empty dict:

```suggestion
    general_fragment = fragments.pop(TDR_FRAGMENT_GENERAL, {})
```

aherbst-broad commented on scripts/convert_mapping_to_schema.py:
```diff
@@ -0,0 +1,117 @@
+    for fragment, fields in fragments.items():
+        schema = render_schema_fragment(fragment, fields, table_name)
+        rendered_json = json.dumps(schema, indent=4)
+        print(os.getcwd())
```
⛏️ Unless this debug output is really needed remove it.

aherbst-broad commented on scripts/convert_mapping_to_schema.py:
```diff
@@ -0,0 +1,117 @@
+        schema = render_schema_fragment(fragment, fields, table_name)
+        rendered_json = json.dumps(schema, indent=4)
+        print(os.getcwd())
+        with open(f"schema/{table_name}_{fragment}.fragment.json", "w") as f:
```
🔧 We should parameterize the output path (ie consume via a cli arg and thread through)

aherbst-broad commented on scripts/convert_mapping_to_schema.py:
```diff
@@ -0,0 +1,117 @@
+            f.write(rendered_json)
+
+    general_schema = render_general_fragment(general_fragment, [f"{table_name}_{key}" for key in fragments.keys()], table_name)
+    with open(f"schema/{table_name}.table.json", "w") as f:
```
🔧 Same comment about parameterizing the output path.

aherbst-broad commented on scripts/convert_mapping_to_schema.py:
```diff
@@ -0,0 +1,117 @@
+        f.write(json.dumps(general_schema, indent=4))
+
+
+def run(table_name: str, mapping_csv: str) -> None:
```
⛏️ This could use a more descriptive name. Perhaps `render_jade_table_schemas`?

aherbst-broad reviewed (changes requested):
Looking good, a couple of points that could use cleanup in the schema generation script.

quazi-h committed:
Housekeeping - descriptions, formatting, syntax, readability changes
New output path argument

aherbst-broad reviewed (approved):
lgtm

quazi-h committed:
Changed the return type of the main transformation function to Eols instead of an option.
Added a check for the willing to complete flag for the general EOLS transformations.
Added a unit test case for a non willing participant.

quazi-h committed:
Merge branch 'master' into qh-arh-dspdc-1808-convert-tool

quazi-broad commented on dap-etl/transformation/src/main/scala/org/broadinstitute/monster/dap/eols/EolsTransformationPipelineBuilder.scala:
```diff
@@ -0,0 +1,63 @@
+    // Group by study ID (record number) and arm (address_year_month)
+    // to get the format: (studyId, arm_id, Iterable((fieldName, Iterable(value))))
+    // (study_id, arm_id, iterable())
+    rawRecords
```
Added a new ticket to address this: [DSPDC-1817](https://broadinstitute.atlassian.net/browse/DSPDC-1817)

quazi-h committed:
Apply suggestions from code review

Co-authored-by: aherbst-broad <74209955+aherbst-broad@users.noreply.github.com>

quazi-h committed:
newline
